gcc + cartesian_product + views::single = unexpected behavior

I have a test code that clang and gcc handle very differently:

```c++
#include <vector>
#include <tuple>
#include <cassert>
#include <range/v3/all.hpp>


int main(int argc, char *argv[])
{
    auto x = std::vector<double>(2, -1.0);

    auto p = ranges::views::cartesian_product(x,
					      ranges::views::single(3.0));
    assert(p.size() == x.size());
    assert(p.end() - p.begin() == 2);

    auto t = ranges::views::transform([](const std::tuple<double, double>& a){
					  return std::get<1>(a);
				      });
    auto v = p | t | ranges::to<std::vector>();

    assert(v.size() == x.size());
    assert(v.back() == 3.0);

}
```
I'm running linux and compiling with clang v.7.0.1 and running works fine, but I get compile errors using gcc 8.3.  The gcc error is:
```
/home/ptb/gitrepos/range-v3/include/range/v3/view/cartesian_product.hpp:320:70: error: could not convert ‘ranges::tuple_transform_fn::operator()(Tup&&, Fun) const [with Tup = const std::tuple<__gnu_cxx::__normal_iterator<double*, std::vector<double, std::allocator<double> > >, const double*>&; Fun = ranges::detail::dereference_fn; decltype (ranges::tuple_transform_fn::impl1(static_cast<Tup&&>(tup), fun, ranges::tuple_indices_t<Tup>{})) = std::tuple<double&, const double&>](((const ranges::cartesian_product_view<ranges::ref_view<std::vector<double, std::allocator<double> > >, ranges::single_view<double> >::cursor<true>*)this)->ranges::cartesian_product_view<ranges::ref_view<std::vector<double, std::allocator<double> > >, ranges::single_view<double> >::cursor<true>::its_, (ranges::detail::dereference_fn{}, ranges::detail::dereference_fn()))’ from ‘std::tuple<double&, const double&>’ to ‘ranges::common_tuple<double&, double&>’
                 return tuple_transform(its_, detail::dereference_fn{});
```

If I use a one element vector instead of `ranges::views::single` then everything works as expected.  I'm using the master branch at commit cabc3039.

--- a/include/range/v3/view/cartesian_product.hpp
+++ b/include/range/v3/view/cartesian_product.hpp
@@ -281,14 +281,14 @@ namespace ranges
                    std::true_type) // common_with
               : cursor(begin_tag{}, view)
             {
-                CPP_assert(common_range<meta::at_c<meta::list<Views...>, 0>>);
+                CPP_assert(common_range<meta::at_c<meta::list<constify_if<Views>...>, 0>>);
                 std::get<0>(its_) = ranges::end(std::get<0>(view->views_));
             }
             cursor(end_tag, constify_if<cartesian_product_view> * view,
                    std::false_type) // !common_with
               : cursor(begin_tag{}, view)
             {
-                using View0 = meta::at_c<meta::list<Views...>, 0>;
+                using View0 = meta::at_c<meta::list<constify_if<Views>...>, 0>;
                 CPP_assert(!common_range<View0> && random_access_range<View0> &&
                            sized_range<View0>);
                 std::get<0>(its_) += ranges::distance(std::get<0>(view->views_));
@@ -308,14 +308,14 @@ namespace ranges
             }
             explicit cursor(end_tag, constify_if<cartesian_product_view> * view)
               : cursor(end_tag{}, view,
-                       meta::bool_<common_range<meta::at_c<meta::list<Views...>, 0>>>{})
+                       meta::bool_<common_range<meta::at_c<meta::list<constify_if<Views>...>, 0>>>{})
             {}
             CPP_template(bool Other)( //
                 requires IsConst_ && (!Other)) cursor(cursor<Other> that)
               : view_(that.view_)
               , its_(std::move(that.its_))
             {}
-            common_tuple<range_reference_t<Views>...> read() const
+            common_tuple<range_reference_t<constify_if<Views>>...> read() const
             {
                 return tuple_transform(its_, detail::dereference_fn{});
             }

--- a/test/view/cartesian_product.cpp
+++ b/test/view/cartesian_product.cpp
@@ -26,7 +26,9 @@
 #include <range/v3/view/indices.hpp>
 #include <range/v3/view/iota.hpp>
 #include <range/v3/view/reverse.hpp>
+#include <range/v3/view/single.hpp>
 #include <range/v3/view/take_exactly.hpp>
+#include <range/v3/view/transform.hpp>
 #include <range/v3/view/filter.hpp>
 #include "../simple_test.hpp"
 #include "../test_utils.hpp"
@@ -266,6 +268,18 @@ void test_bug_1279()
     }
 }
 
+void test_bug_1296()
+{
+    // https://github.com/ericniebler/range-v3/issues/1296
+    auto v = ranges::views::cartesian_product(ranges::views::single(2.0))
+        | ranges::views::transform([](std::tuple<double> a) {
+            return std::get<0>(a);
+        });
+
+    CHECK(ranges::size(v) == 1u);
+    CHECK(*ranges::begin(v) == 2.0);
+}
+
 int main()
 {
     int some_ints[] = {0,1,2,3};
@@ -321,6 +335,7 @@ int main()
     test_bug_978();
     test_bug_1269();
     test_bug_1279();
+    test_bug_1296();
 
     return test_result();
 }

--- a/test/view/cartesian_product.cpp
+++ b/test/view/cartesian_product.cpp
@@ -271,13 +271,13 @@ void test_bug_1279()
 void test_bug_1296()
 {
     // https://github.com/ericniebler/range-v3/issues/1296
-    auto v = ranges::views::cartesian_product(ranges::views::single(2.0))
-        | ranges::views::transform([](std::tuple<double> a) {
+    auto v = ranges::views::cartesian_product(ranges::views::single(42))
+        | ranges::views::transform([](std::tuple<int> a) {
             return std::get<0>(a);
         });
 
     CHECK(ranges::size(v) == 1u);
-    CHECK(*ranges::begin(v) == 2.0);
+    CHECK(*ranges::begin(v) == 42);
 }
 
 int main()

